File: 33_Search_in_Rotated_Sorted_Array/solution.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        /*
            Input: vector of integers nums initially sorted in ascending order with distinct values and rotated, integer target
            Output: integer the index of target if in nums else -1

            Rotated: shift left

            Constraints: write in O(logn)

            Binary search in a sorted rotated array

            Sample test case
            1.) nums = [4,5,6,7,0,1,2], target = 0, output: 4
                - if array is not rotated -> binary search as per normal
                - if array is rotated, there exist a partition where both sides are sorted themselves
                - We can narrow down which half target will be with the following conditions
                    - if leftmost element <= middle
                        - if target resides in between leftmost and middle:
                            - search left
                        - else search right
                    - if leftmost element > middle
                        - if target element is smaller than middle yet it is smaller than right bound
                            - search right
                        - else search left
                
        */

        int l = 0;
        int r = nums.size()-1;
        int mid;

        while (l<=r) {
            mid = l + (r-l)/2;
            if (nums[mid] == target){
                return mid;
            }
            if (nums[l] <= nums[mid]) {
                if (nums[l] <= target && target <= nums[mid]){
                    r = mid-1;
                }else{
                    l = mid+1;
                }
            }else{
                if (nums[mid] <= target && target <= nums[r]){
                    l = mid+1;
                }else{
                    r = mid-1;
                }
            }
            
        }
        return -1;
        
    }
};

File: 33_Search_in_Rotated_Sorted_Array/solution.java
class Solution {
    public int search(int[] nums, int target) {
        /*
            Input: integer array nums, integer target
            Output: index of target if present in nums else -1

            Observation
            - rotated array consists of two sorted arrays
            - we want to determine first in which sorted array is target located in
                - We can do it through binary search
                - first determine that current array is rotated
                    - if nums[left] > nums[right]
                - if nums[mid] is target, return mid (got it by chance or with subsequent searches)
                - if it is rotated 
                    
                    - we check with nums[mid] if it is < than nums[right] (mid is located on the right side arr)
                        - if target lies between mid and right, we can shift left to mid + 1
                        - else shift right to mid (narrow search range to left side arr since target lies there)
                    - else it means mid is located in the left side arr
                        - if target lies between mid and left, shift right to mid-1
                        - else shift left to mid
                - else we have already narrowed the "right" array
                    - if nums[mid] < target
                        - search right
                    - else
                        - search left
        */
        int left=0;
        int right=nums.length-1;
        int mid;

        while (left <= right) {
            mid = left + (right-left)/2;
            if (nums[mid] == target) return mid;
            if (nums[left] > nums[right]) {
                if (nums[mid] < nums[right])
                    if (target >= nums[mid] && target <= nums[right]) 
                        left = mid+1;
                    else
                        right = mid-1;
                else
                    if (target >= nums[left] && target <= nums[mid]) 
                        right = mid-1;
                    else
                        left = mid+1;
                       
            }else{
                if (nums[mid] < target) {
                    left = mid+1;
                }else{
                    right = mid-1;
                }
            }
        }
        return -1;
    }
}
